pcap-linux.c

```diff
@@ -3677,54 +3246,20 @@ init_tpacket(pcap_t *handle, int version, const char *version_str)
 
 	return 0;
 }
-#endif /* defined HAVE_TPACKET2 || defined HAVE_TPACKET3 */
-
-/*
- * If the instruction set for which we're compiling has both 32-bit
- * and 64-bit versions, and Linux support for the 64-bit version
- * predates TPACKET_V2, define ISA_64_BIT as the .machine value
- * you get from uname() for the 64-bit version.  Otherwise, leave
- * it undefined.  (This includes ARM, which has a 64-bit version,
- * but Linux support for it appeared well after TPACKET_V2 support
- * did, so there should never be a case where 32-bit ARM code is
- * running o a 64-bit kernel that only supports TPACKET_V1.)
- *
- * If we've omitted your favorite such architecture, please contribute
- * a patch.  (No patch is needed for architectures that are 32-bit-only
- * or for which Linux has no support for 32-bit userland - or for which,
- * as noted, 64-bit support appeared in Linux after TPACKET_V2 support
- * did.)
- */
-#if defined(__i386__)
-#define ISA_64_BIT	"x86_64"
-#elif defined(__ppc__)
-#define ISA_64_BIT	"ppc64"
-#elif defined(__sparc__)
-#define ISA_64_BIT	"sparc64"
-#elif defined(__s390__)
-#define ISA_64_BIT	"s390x"
-#elif defined(__mips__)
-#define ISA_64_BIT	"mips64"
-#elif defined(__hppa__)
-#define ISA_64_BIT	"parisc64"
-#endif
 
 /*
  * Attempt to set the socket to version 3 of the memory-mapped header and,
  * if that fails because version 3 isn't supported, attempt to fall
- * back to version 2.  If version 2 isn't supported, just leave it at
- * version 1.
+ * back to version 2.  If version 2 isn't supported, just fail.
  *
- * Return 1 if we succeed or if we fail because neither version 2 nor 3 is
- * supported; return -1 on any other error, and set handle->errbuf.
+ * Return 1 if we succeed or if we fail because is not supported;
```

Comment:
This only returns 1 if it succeeds, at least as I read it; -1 is returned if we fail for any reason.

Also, shouldn't `init_tpacket()` return 0 on success and -1 for any failure?  ENOPROTOOPT from `getsockopt(..., SOL_PACKET, PACKET_HDRLEN, ...)` means "we don't support memory-mapped capture at all, and EINVAL means "we don't support anything other than TPACKET_V1, so `init_tpacket()` should return -1 with an error message saying "your kernel is too old to support TPACKET_V2 memory-mapped capture" (I'm guessing that, by the time TPACKET_V2 was added, you can't un-configure PACKET_MMAP - if not, perhaps the error should distinguish between "kernel is too old" and "kernel isn't too old, but you have to configure PACKET_MMAP).

Given that, `prepare_tpacket_socket()` should treat -1 as an error and anything else as success.

Reply:
1. Fixed on my local repo.
2. Yes, you're correct, it's used as if it returns 0 on success and -1 on error.
3. I'm not sure about the PACKET_MMAP configuration option, but on kernel version I though of throwing an error on initialization on unsupported kernels instead.
4. Makes sense, I'll take a look.

On next review, could you take special attention in the statistics changes? I think those are harder to properly test, so the extra caution will prove useful.

Reply:
OK, I *think* I addressed all of these.